Rework concurrency

tbg opened the pull request:
This commit basically rewrites `stress`. The reason I touched it was
because I wanted better behavior for `-maxtime` in the sense that at
least one iteration should complete before declaring success. This
was not previously the case, for example

    stress -maxtime 1s /bin/bash -c "sleep 5; false"

would previously return with exit status zero (and zero iterations run).
This behavior renders `-maxtime` unusable for timeboxing stress
invocations effectively, since the executable may just hang forever.
This could be partially mitigated via the `-timeout` flag, but that's
not a reasonable solution. For example, in CockroachDB nightly stress,
we invoke various packages which can take ~10 minutes to run, and so
we can't set `-timeout` smaller than, say, 20 minutes. However, by
blindly specifying `-maxruns 100` some packages may be stressed for days
at a time, which isn't what we want to do. Supplying `-maxtime 1h`
previously did not give us confidence that the package was run to
completion at all.

With this commit, it's perfectly reasonable to specify `-maxtime 1h` to
mean "run at least once, but stop adding new invocations after an hour
has passed and let things settle down".

While rewriting this I added relatively comprehensive tests. They don't
quite cover 100% of functionality (for example outputting to temp dir)
but they give more confidence that everything works than I had about
the old code.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/stress/12)
<!-- Reviewable:end -->

tbg committed:
Rework concurrency

This commit basically rewrites `stress`. The reason I touched it was
because I wanted better behavior for `-maxtime` in the sense that at
least one iteration should complete before declaring success. This
was not previously the case, for example

    stress -maxtime 1s /bin/bash -c "sleep 5; false"

would previously return with exit status zero (and zero iterations run).
This behavior renders `-maxtime` unusable for timeboxing stress
invocations effectively, since the executable may just hang forever.
This could be partially mitigated via the `-timeout` flag, but that's
not a reasonable solution. For example, in CockroachDB nightly stress,
we invoke various packages which can take ~10 minutes to run, and so
we can't set `-timeout` smaller than, say, 20 minutes. However, by
blindly specifying `-maxruns 100` some packages may be stressed for days
at a time, which isn't what we want to do. Supplying `-maxtime 1h`
previously did not give us confidence that the package was run to
completion at all.

With this commit, it's perfectly reasonable to specify `-maxtime 1h` to
mean "run at least once, but stop adding new invocations after an hour
has passed and let things settle down".

While rewriting this I added relatively comprehensive tests. They don't
quite cover 100% of functionality (for example outputting to temp dir)
but they give more confidence that everything works than I had about
the old code.
